src/server/tran_server.cpp

```diff
@@ -384,15 +390,55 @@ tran_server::connection_handler::connection_handler (cubcomm::channel &&chn, tra
   m_conn->start ();
 }
 
+tran_server::connection_handler::~connection_handler ()
+{
+  send_disconnect_request ();
+  m_conn->stop_incoming_communication_thread ();
+  m_conn->stop_outgoing_communication_thread ();
+
+  er_log_debug (ARG_FILE_LINE, "Transaction server is disconnected from the page server with channel id: %s \n",
+		get_channel_id ().c_str ());
+}
+
 tran_server::connection_handler::request_handlers_map_t
 tran_server::connection_handler::get_request_handlers ()
 {
   // Insert handlers specific to all transaction servers here.
   // For now, there are no such handlers; return an empty map.
   std::map<page_to_tran_request, page_server_conn_t::incoming_request_handler_t> handlers_map;
+
+  auto disconnect_request_handler = std::bind (&tran_server::connection_handler::receive_disconnect_request, this,
+				    std::placeholders::_1);
+
+  handlers_map.insert (std::make_pair (page_to_tran_request::SEND_DISCONNECT_REQUEST_MSG, disconnect_request_handler));
+
   return handlers_map;
 }
 
+void
+tran_server::connection_handler::receive_disconnect_request (page_server_conn_t::sequenced_payload &a_ip)
+{
+  m_conn->stop_response_broker (); // wake up threads waiting for a response and tell them it won't be served.
+
+  {
+    std::lock_guard<std::shared_mutex> lk_guard (m_ts.m_page_server_conn_vec_mtx);
+    auto &conn_vec = m_ts.m_page_server_conn_vec;
+
+    auto it = conn_vec.begin ();
+    for (; it != conn_vec.end (); it++)
+      {
+	if (it->get () == this)
+	  {
+	    m_ts.m_async_disconnect_handler.disconnect (std::move (*it));
+	    assert (*it == nullptr);
+	    conn_vec.erase (it);
+	    break;
+	  }
+      }
+  }
```

Comment:
Nit.
This should be a function on `tran_server` - `m_ts.disconnect_me(this)` - since it works will all the objects that are member of `tran_server`.
```suggestion
  {
    std::lock_guard<std::shared_mutex> lk_guard (m_ts.m_page_server_conn_vec_mtx);
    auto &conn_vec = m_ts.m_page_server_conn_vec;

    auto it = conn_vec.begin ();
    for (; it != conn_vec.end (); it++)
      {
	if (it->get () == this)
	  {
	    m_ts.m_async_disconnect_handler.disconnect (std::move (*it));
	    assert (*it == nullptr);
	    conn_vec.erase (it);
	    break;
	  }
      }
  }
```

Reply:
Thank you for the pointer. https://github.com/CUBRID/cubrid/pull/4135/commits/f24b2e1ccb42f59c7ffb869482735b33a5ba6941